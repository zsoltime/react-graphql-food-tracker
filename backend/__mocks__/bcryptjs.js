'use strict';

const bcryptjs = {};

let result = true;

bcryptjs.setMockResult = (newMockResult) => {
  result = newMockResult;
};

bcryptjs.compare = () => Promise.resolve(result);

module.exports = bcryptjs;
