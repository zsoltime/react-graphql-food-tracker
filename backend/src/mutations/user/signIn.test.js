'use strict';

const signIn = require('./signIn');

jest.mock('bcryptjs');

// There are two describes in thus file as I wanted
// to mock bcrypt.compare with different results
describe('signIn mutation happy path', () => {
  beforeEach(() => {
    // eslint-disable-next-line global-require
    require('bcryptjs').setMockResult(true);
  });

  it('signs user in', async () => {
    const args = { email: 'test@email.io', password: 'test1234' };
    const fakeUser = {
      id: 'testId',
      email: 'test@email.io',
      pendingEmail: null,
      permissions: ['USER'],
      profile: null,
    };
    const mockedContext = {
      db: {
        query: {
          user: jest
            .fn()
            .mockImplementation(() => Promise.resolve(fakeUser)),
        },
      },
      response: {
        cookie: jest.fn(),
      },
    };

    const user = await signIn(null, args, mockedContext);

    expect(mockedContext.response.cookie).toHaveBeenCalled();
    expect(
      mockedContext.response.cookie.mock.calls[0]
    ).toMatchSnapshot();
    expect(user).toMatchSnapshot();
  });
});

describe('signIn mutation', () => {
  beforeEach(() => {
    // eslint-disable-next-line global-require
    require('bcryptjs').setMockResult(false);
  });

  it('throws error if user is not found in database', () => {
    const args = { email: 'zsolti@me.com', password: 'test1234' };
    const mockedContext = {
      db: {
        query: {
          user: jest.fn().mockImplementation(() => null),
        },
      },
    };

    expect(signIn(null, args, mockedContext)).rejects.toThrowError(
      'Email or password is incorrect'
    );

    expect(mockedContext.db.query.user).toHaveBeenCalledWith({
      where: { email: args.email },
    });
  });

  it('throws error if password is incorrect', () => {
    const args = { email: 'zsolti@me.com', password: 'test1234' };
    const mockedContext = {
      db: {
        query: {
          user: jest
            .fn()
            .mockImplementation(() => Promise.resolve({ password: 'test' })),
        },
      },
    };

    expect(signIn(null, args, mockedContext)).rejects.toThrowError(
      'Email or password is incorrect'
    );
  });
});
