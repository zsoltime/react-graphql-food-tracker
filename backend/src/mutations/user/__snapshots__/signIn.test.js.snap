// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`signIn mutation happy path signs user in 1`] = `
Array [
  "token",
  "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VySWQiOiJ0ZXN0SWQiLCJpYXQiOjE1NDMyNDE2MDZ9.pvuxlFHMguOedl_DDC8bYZqTW7Bjku-xTiKDEryuiws",
  Object {
    "httpOnly": true,
    "maxAge": 7776000000,
    "secure": false,
  },
]
`;

exports[`signIn mutation happy path signs user in 2`] = `
Object {
  "email": "test@email.io",
  "id": "testId",
  "pendingEmail": null,
  "permissions": Array [
    "USER",
  ],
  "profile": null,
}
`;
